File: machine/cpp/lua/module_path.cpp
#include "lua/module_path.h"

#include <algorithm>

namespace bmsx {

namespace {

bool startsWith(std::string_view value, std::string_view prefix) {
	return value.size() >= prefix.size() && value.compare(0, prefix.size(), prefix) == 0;
}

bool hasLuaExtension(std::string_view candidate) {
	if (candidate.size() < 4) return false;
	const size_t dotIndex = candidate.size() - 4;
	return candidate[dotIndex] == '.'
		&& (candidate[dotIndex + 1] == 'l' || candidate[dotIndex + 1] == 'L')
		&& (candidate[dotIndex + 2] == 'u' || candidate[dotIndex + 2] == 'U')
		&& (candidate[dotIndex + 3] == 'a' || candidate[dotIndex + 3] == 'A');
}

} // namespace

std::string stripLuaExtension(std::string_view candidate) {
	if (hasLuaExtension(candidate)) candidate.remove_suffix(4);
	return std::string(candidate);
}

std::string toLuaModulePath(std::string_view sourcePath) {
	static constexpr std::string_view CART_SOURCE_PREFIX = "carts/";
	static constexpr std::string_view MODULE_PATH_SOURCE_PREFIXES[] = {
		"machine/firmware/res/",
		"machine/firmware/",
		"res/",
	};
	std::string path = stripLuaExtension(sourcePath);
	std::replace(path.begin(), path.end(), '\\', '/');
	std::string_view modulePath = path;
	if (startsWith(path, CART_SOURCE_PREFIX)) {
		modulePath.remove_prefix(CART_SOURCE_PREFIX.size());
		const size_t cartNameEnd = modulePath.find('/');
		if (cartNameEnd != std::string_view::npos) modulePath.remove_prefix(cartNameEnd + 1);
	} else {
		for (const std::string_view prefix : MODULE_PATH_SOURCE_PREFIXES) {
			if (startsWith(path, prefix)) {
				modulePath.remove_prefix(prefix.size());
				break;
			}
		}
	}
	return std::string(modulePath);
}

} // namespace bmsx

File: toolchain/ts/lua/module_path.ts
export function stripLuaExtension(candidate: string): string {
	const lower = candidate.toLowerCase();
	if (lower.endsWith('.lua')) {
		return candidate.slice(0, candidate.length - 4);
	}
	return candidate;
}

const CART_SOURCE_PREFIX = 'carts/';
const BIOS_RESOURCE_SOURCE_PREFIX = 'machine/bios/res/';
const BIOS_SOURCE_PREFIX = 'machine/bios/';
const RESOURCE_SOURCE_PREFIX = 'res/';
const MODULE_PATH_SOURCE_PREFIXES = [
	BIOS_RESOURCE_SOURCE_PREFIX,
	BIOS_SOURCE_PREFIX,
	RESOURCE_SOURCE_PREFIX,
];

export function toLuaModulePath(sourcePath: string): string {
	const path = stripLuaExtension(sourcePath.includes('\\') ? sourcePath.replace(/\\/g, '/') : sourcePath);
	if (path.startsWith(CART_SOURCE_PREFIX)) {
		const cartRelative = path.substring(CART_SOURCE_PREFIX.length);
		const cartNameEnd = cartRelative.indexOf('/');
		return cartNameEnd >= 0 ? cartRelative.substring(cartNameEnd + 1) : cartRelative;
	}
	for (let index = 0; index < MODULE_PATH_SOURCE_PREFIXES.length; index += 1) {
		const prefix = MODULE_PATH_SOURCE_PREFIXES[index];
		if (path.startsWith(prefix)) {
			return path.substring(prefix.length);
		}
	}
	return path;
}

export const buildModuleExportPathKey = (path: ReadonlyArray<string>): string =>
	path.join('.');

export const appendModuleExportPathKey = (base: string, key: string): string =>
	base.length === 0 ? key : `${base}.${key}`;

const sanitizeModuleSlotSegment = (value: string): string =>
	value.replace(/[^A-Za-z0-9_]/g, '_');

export function buildModuleExportSlotName(
	modulePath: string,
	exportPath: ReadonlyArray<string>,
): string {
	const compactPath = toLuaModulePath(modulePath);
	const parts = compactPath.split('/');
	let out = '';
	for (let index = 0; index < parts.length; index += 1) {
		const part = parts[index];
		if (part.length > 0) {
			out += out.length === 0
				? sanitizeModuleSlotSegment(part)
				: `__${sanitizeModuleSlotSegment(part)}`;
		}
	}
	if (out.length === 0) {
		out = sanitizeModuleSlotSegment(compactPath);
	}
	for (let index = 0; index < exportPath.length; index += 1) {
		out += `__${sanitizeModuleSlotSegment(exportPath[index])}`;
	}
	return out;
}
